### PR TITLE
Half up rounding

### PR DESCRIPTION
**Pull request by hbe72**
Towards infinity rounding for fixed_point datatypes with only values midway between two representable values move in the given direction. Other values move to the nearest representable value. The tag is named as "towards_infinity_rounding_tag".

### CODE REVIEW

**Comment by johnmcfarlane** [comment]
@hbe72 are you able to add johnmcfarlane/cnl as a remote and to push to that?

```sh
git remote add upstream git@github.com:johnmcfarlane/cnl
git fetch upstream
git push -u upstream half_up_rounding
```

Happy to push this for you but for next time, if you're able to push directly to a feature branch in *johnmcfarlane/cnl* then CI should work for you.

File: src/test/_impl/rounding/convert_operator.cpp [hunk 1]
```diff
@@ -82,3 +90,47 @@ namespace test_convert_nearest_rounding_scaled_integer
...
+            "cnl::convert<towards_infinity_rounding_tag, int, float>");
+
+    static_assert(
+            cnl::_impl::identical(1, cnl::convert<cnl::towards_infinity_rounding_tag, int, float>(0.5F)),
+            "cnl::convert<towards_infinity_rounding_tag, int, float>");
+
+    static_assert(
+            cnl::_impl::identical(1, cnl::convert<cnl::towards_infinity_rounding_tag, int, float>(0.725F)),
+            "cnl::convert<towards_infinity_rounding_tag, int, float>");
+
```
Comment (johnmcfarlane) [inline src/test/_impl/rounding/convert_operator.cpp]: Suggested test for completeness:

```c++
    static_assert(
            cnl::_impl::identical(-1, cnl::convert<cnl::towards_infinity_rounding_tag, int, float>(-0.125F)),
            "cnl::convert<towards_infinity_rounding_tag, int, float>");
```

**Comment by johnmcfarlane** [comment]
I made a couple of comments in the code. LMK if you don't see them.

I think I may have misunderstood the intent of this mode. I'm expecting something which behaves the same as [`ceil`](https://en.cppreference.com/w/cpp/numeric/math/ceil) for +ve values and [`floor`](https://en.cppreference.com/w/cpp/numeric/math/floor) for -ve values.

Also, is this not the opposite of what you're aiming for? Don't you want the same behavior as right-shift? If so, I would think you want to round toward -ve infinity.

File: src/test/_impl/rounding/convert_operator.cpp (hunk 1 above)
Comment (hbe72) [reply src/test/_impl/rounding/convert_operator.cpp]: This should round to 0
    static_assert(
            cnl::_impl::identical(0, cnl::convert<cnl::towards_infinity_rounding_tag, int, float>(-0.125F)),
            "cnl::convert<towards_infinity_rounding_tag, int, float>");

**Comment by johnmcfarlane** [comment]
I see where my confusion arose. Previous discussion, #405, refers to a fixed-point behavior which is optimised for use of shift operator. That's what I assumed you were talking about. I thought perhaps you were describing this as 'add 0.5 then round to nearest' but you're actually looking for 'round to nearest or to +ve infinity where there is no nearest'. 

So sorry but `towards_infinity_rounding_tag` _isn't right after all_.

I previously claimed that your desired rounding mode was `round_to_infinity` in [P0105R1](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0105r1.html#RoundMode) but now I think it's actually `tie_to_pos_inf`. Perhaps you can confirm that and consider this name. Alternatively, 'round half up' or 'half round up' might be better. 

I'm frankly indifferent to all of them. You need to know what `tie` and `all` are before Lawrence's enumerators make sense. Alternatively, I don't see a citation for terminology involving 'half' and it doesn't leap out as obvious. In the end, I think this is quite a niche behavior: look at all of P0105R1's options. It's a lot of obscure customization!

_Please do the honors and pick one._ :)

Longer term, I think whatever you chose is likely to become an alias for a parameterised tag of the form:

```c++
namespace cnl {
using half_up_rounding_tag = nearest_rounding_tag<positive_inf>;
}
```

One more correction: `nearest_rounding` isn't as well defined as you think. It's more like `round_to_nearest_but_dont_care_how`. In that sense, it's a bit more like the "I'm not picky" option from P0106R1!

**Comment by hbe72** [comment]
I think now we are on the same page ;). I'm exactly looking for 'round to nearest or to +ve infinity where there is no nearest'.  I just did not have the correct (exact) vocabulary to describe it.

I had previously missed the end of the P0105R1, the enumerations, and now the rest of the paper makes more sense. tie_to_pos_inf is the correct choice. Is it then tie_to_pos_inf_rounding_tag?

**Comment by johnmcfarlane** [comment]
Yes, that's fine.

**Comment by hbe72** [comment]
Duplicate. Closed.

**Pull request closed without merging**